boot/freeldr/freeldr/lib/fs/iso.c

```diff
@@ -140,116 +171,151 @@ static ARC_STATUS IsoBufferDirectory(ULONG DeviceId, ULONG DirectoryStartSector,
+    Volume = IsoVolumes[DeviceId];
 
-    //
-    // Read the Primary Volume Descriptor
-    //
-    Position.HighPart = 0;
-    Position.LowPart = 16 * SECTORSIZE;
-    Status = ArcSeek(DeviceId, &Position, SeekAbsolute);
-    if (Status != ESUCCESS)
-        return Status;
-    Status = ArcRead(DeviceId, Pvd, SECTORSIZE, &Count);
-    if (Status != ESUCCESS || Count < sizeof(PVD))
-        return EIO;
+    /*
+     * Extract the file name
+     * '\test.ini' --> 'test.ini'
+     * '\folder\app.exe' --> 'app.exe'
+     */
+    FileNameStr = IsoLastPathSeparator(FileName) + 1;
 
-    DirectorySector = Pvd->RootDirRecord.ExtentLocationL;
-    DirectoryLength = Pvd->RootDirRecord.DataLengthL;
+    /*
+     * Extract the directory path, including trailing slash
+     * '\test.ini' --> '\'
+     * '\folder\app.exe' --> '\folder\'
+     */
+    DirectoryPathLength = FileNameStr - FileName;
 
-    /* Skip leading path separator, if any */
-    if (*FileName == '\\' || *FileName == '/')
-        ++FileName;
-    //
-    // Figure out how many sub-directories we are nested in
-    //
-    NumberOfPathParts = FsGetNumPathParts(FileName);
+    /* See if this directory has been buffered before */
+    if ((DirectoryPathLength == Volume->DirectoryPathLength) &&
+        RtlEqualMemory(FileName, Volume->DirectoryPath, DirectoryPathLength))
+    {
+        PathPart = (CHAR*)FileNameStr;
+        DoFullLookup = FALSE;
+
+        DirectoryBuffer = Volume->DirectoryBuffer;
+        DirectoryLength = Volume->DirectoryLength;
+
+        NumberOfPathParts = 1;
+    }
+    else
+    {
+        PathPart = PathBuffer;
+        DoFullLookup = TRUE;
+
+        DirectorySector = Volume->PvdDirectorySector;
+        DirectoryLength = Volume->PvdDirectoryLength;
+
+        /* Skip leading path separator, if any */
+        if (*FileName == '\\' || *FileName == '/')
+            ++FileName;
+
+        /* Figure out how many sub-directories we are nested in */
+        NumberOfPathParts = FsGetNumPathParts(FileName);
+    }
 
     //
     // Loop once for each part
     //
     for (i=0; i<NumberOfPathParts; i++)
     {
-        //
-        // Get first path part
-        //
-        FsGetFirstNameFromPath(PathPart, FileName);
-
-        //
-        // Advance to the next part of the path
-        //
-        for (; (*FileName != '\\') && (*FileName != '/') && (*FileName != '\0'); FileName++)
+        if (DoFullLookup)
         {
-        }
-        FileName++;
+            /* Get first path part */
+            FsGetFirstNameFromPath(PathPart, FileName);
 
-        //
-        // Buffer the directory contents
-        //
-        Status = IsoBufferDirectory(DeviceId, DirectorySector, DirectoryLength, &DirectoryBuffer);
-        if (Status != ESUCCESS)
-            return Status;
+            /* Advance to the next part of the path */
+            for (; (*FileName != '\\') && (*FileName != '/') && (*FileName != '\0'); FileName++)
```

Comment:
Note that `'\0'` should (logically) be tested first, and could use `ANSI_NULL`,
but it's not worth changing it, unless doing that in the 3 other files too.
https://git.reactos.org/?p=reactos.git&a=search&h=HEAD&st=grep&s=for+%28%3B+%28*FileName+%21%3D+%27%5C%5C%27%29+%26%26+%28*FileName+%21%3D+%27%2F%27%29+%26%26+%28*FileName+%21%3D+%27%5C0%27%29%3B+FileName%2B%2B%29

Reply:
That is a good point.